### PR TITLE
Java records article

### PR DESCRIPTION
**Pull request by sagaofsilence**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [x] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [x] the article explains *why* and not only *how*
- [x] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [x] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [x] the text is conversational ("we" instead of "you", easy language)
- [x] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [x] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [x] the text uses active voice instead of passive voice
- [x] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [x] file names and variables are highlighted as `code`

## Structure
- [x] the article has a short intro before the first headline
- [x] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [x] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [x] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [x] code examples are introduced with a ":" and explained below the code example
- [x] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

**Comment by sagaofsilence** [comment]
@thombergs please review the article on Java records.

**Comment by Maiklins** [comment]
Overall nice article. I left a few comments in the PR. A few things that you might want to add to the article are:
- records extend java.lang.Record, are final, and cannot be extended
- a quick note about Lombok and Kotlin

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 1]
```diff
@@ -0,0 +1,511 @@
+---
+authors: [sagaofsilence]
+title: "Use Cases for Java Records"
+categories: ["Java"]
+date: 2023-12-03 00:00:00 +0530
+excerpt: "Get familiar with the Java records."
+image: images/stock/0069-testcontainers-1200x628.jpg
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:7]: OK. Done.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+}
+```
+
+### Boilerplate Fatigue
+Writing repetitive boilerplate code was not only time-consuming but also error-prone. Any change in the class structure meant manually updating multiple methods, leading to maintenance nightmares.
+
+### Immutability Hassles
+Ensuring immutability, a key aspect of data integrity, demanded additional effort. We had to meticulously design classes to make them immutable, often leading to complex and convoluted code.
+
+## How Does Java Records Address These Issues?
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:135]: Sure. Done.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+
+### Conciseness Redefined
+   Java records introduce a concise syntax that automates the generation of essential methods. They eliminate the need for boilerplate code, allowing us to focus on the core logic of their applications.
+
+### Immutability by Default
+   Records are inherently immutable, meaning once created, their state cannot be changed. This immutability is a fundamental property, ensuring data integrity without us having to enforce it manually.
+
+### Automatic Method Generation
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:143]: ### Simplicity
One of the key advantages of records is their ability to create simple data objects without the need for writing extensive boilerplate code. With records, we can define the structure of our data concisely and straightforwardly.

### Immutability by Default
Java records are immutable by default, meaning that once we create a record, we cannot modify its fields. This immutability ensures data integrity and makes it easier to reason about the state of the objects.

The Simplicity section makes a different point compared to "Immutability by Default".
Am I missing anything here?

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+A record acquires these members automatically:
+- A private final field for each of its components
+- A public read accessor method for each component with the same name and type of the component; in this example, these methods are `Person::name()` and `Person::age()`
+- A public constructor whose signature is derived from the record components list. The constructor initializes each private field from the corresponding argument.
+- Implementations of the `equals()` and `hashCode()` methods, which specify that two records are equal if they are of the same type and their corresponding record components are equal
+- An implementation of the `toString()` method that includes the string representation of all the record's components, with their names
+
+Here's a breakdown of the essential components.
+
+### Record Keyword and Components
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:172]: Removed section '### Record Keyword and Components'

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+
+Address address = new Address("City", List.copyOf(streets));
+streets.add("Street 3");
+
+System.out.println(address.streets()); // Output: [Street 1, Street 2]
+```
+
+Let's explore best practices for using Java records effectively in projects, guidelines for maintaining code consistency and readability, and practical applications and conclusions.
+
+## Best Practices for Using Java Records
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:459]: Done. Kept the heading for each best practice and added text for it.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+
+### Next Steps
+- **Apply the Concepts Learned:** Do not just stop at understanding — implement records in projects. Create immutable, clean, and readable data classes.
+- **Experiment and Refine:** Play with different scenarios and refine understanding. Experimentation leads to mastery.
+- **Share and Collaborate:** Share knowledge with peers, and collaborate on projects to learn from real-world use cases.
+
+### Further Reading and Resources
+- **Official Java Documentation:** Dive deep into the [official documentation](https://docs.oracle.com/en/java/javase/14/index.html) to explore advanced topics and nuances.
+
+### Keep the Learning Journey Alive
```
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:500]: OK. Removed the encouraging generic text.

**Comment by sagaofsilence** [comment]
Changes done. Only one change is pending. https://github.com/reflectoring/reflectoring.github.io/pull/547#discussion_r1476599758

Will fix it once you clarify my question.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maiklins) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md]: What I meant is that the section 

> **Automatic Method Generation**
>    **Records automatically generate methods like `equals()`, `hashCode()`, and `toString()`.** This automation simplifies the development process, making classes more maintainable and reducing the chances of human error.
> 

was very similar to:

> **Why Records?**
> Traditionally, Java classes representing data structures contained repetitive code for **methods such as equals(), hashCode(), toString(), getter/setter methods, and public constructors.** This resulted in bloated classes, making the codebase more difficult to read, understand, maintain, and extend. Java records were introduced as a solution, **simplifying the creation of data-centric classes and addressing the issues of verbosity and redundancy**.
> 
> There are several benefits to using Java records:
> 
> **Simplicity**
> One of the key advantages of records is their ability to create simple data objects without the need for writing extens ive boilerplate code. With records, we can define the structure of our data concisely and straightforwardly.

Maybe not exactly the same, but quite repetitive while reading. However, as you removed the whole section, I think we are good. :)

**Comment by Maiklins** [comment]
> Overall nice article. I left a few comments in the PR. A few things that you might want to add to the article are:
> 
> * records extend java.lang.Record, are final, and cannot be extended
> * a quick note about Lombok and Kotlin

Nice to have the comparison with Kotlin and Lombok in an info box.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+streets.add("Street 3");
+
+System.out.println(address.streets()); // Output: [Street 1, Street 2]
+```
+
+Let's explore best practices for using Java records effectively in projects, guidelines for maintaining code consistency and readability, and practical applications and conclusions.
+
+## Best Practices for Using Java Records
+
+### Choose Appropriate Use Cases
```
Comment (Maiklins) [inline content/blog/2023/2023-12-03-use-cases-for-java-records.md:451]: Let's name a few examples. A nice use case for records are request bodies in a REST API, records read from a database, and DTOs in general. Some code examples for every beast practise would make the article more hands-on.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+### Avoid Business Logic in Records
+Steer clear of embedding business logic within records. Maintain the focus of records on data representation and delegate business logic to separate classes. This approach enhances maintainability and adheres to the principle of separation of concerns.
+
+### Maintain Readability
+To maintain readability, adhere to consistent naming conventions for both record classes and their components. Meaningful names for records and components contribute to code clarity, making it easier for developers to understand the structure and purpose of the classes.
+
+### Document with Comments
+Documenting records with comments is a good practice, particularly for complex or non-intuitive logic within records. Comments provide insights into the rationale behind certain decisions or implementations. Additionally, document any assumptions or constraints associated with record components to facilitate a better understanding of their usage.
```
Comment (Maiklins) [inline content/blog/2023/2023-12-03-use-cases-for-java-records.md:467]: When I read this I got a bit confused. In general, we should avoid adding comments to code as much as possible (okay, that's a bit controversial, of course, but my opinion). But more importantly, this section is a contradiction to what we say earlier about avoiding overcomplication and avoiding business logic. Let's remove this section? What do you think?

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+
+## Mastering Java Records
+
+### Recap of Key Concepts
```
Comment (Maiklins) [inline content/blog/2023/2023-12-03-use-cases-for-java-records.md:471]: Let's also some full sentences for each of these sections.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md]: OK.. Will not touch this section then.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 8 above)
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:467]: Sure. I would remove the mention of *complex or non-intuitive logic* in it. As we do not recommend records to implement complex things. I believe we should document implementation details, assumptions and constraints or limitations. I have come across many such cases. Simple things like, default values, nullable or not, exceptions thrown or consumed, simple condition checks. 

Here is a proposed copy:
Documenting records with comments is a good practice. Comments provide insights into the rationale behind certain decisions or implementations. Additionally, document any assumptions or constraints associated with record components to facilitate a better understanding of their usage.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 9 above)
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:471]: ok

**Comment by sagaofsilence** [comment]
Updated the article as per comments above.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 8 above)
Comment (Maiklins) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md]: I'd remove it as it's a very general statement not especially related to records. If we keep this section, the readers mind might deviate away from the code topic (records) to somethings else, namely the use of comments in code in general. So I think it's better to skip this part.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 7 above)
Comment (Maiklins) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:451]: I think you missed this comment. ;-)

Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:451]: OK.. Will try to come up with appropriate examples

Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:451]: Added 2 examples for each best practice.

**Comment by sagaofsilence** [comment]
Done the requested changes.

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+- Use Case 2: Sensor Readings: Use records to represent sensor readings in an Internet of Things (IoT) application. Keep records focused on data representation, and delegate any complex processing or analysis of sensor data to dedicated classes.
+
+### Maintain Readability
+To maintain readability, adhere to consistent naming conventions for both record classes and their components. Meaningful names for records and components contribute to code clarity, making it easier for developers to understand the structure and purpose of the classes.
+
+- Use Case 1: User Preferences: Apply records to store user preferences in a system. Consistent naming conventions for record classes and components enhance code readability, making it easier for developers to understand and modify user preference-related code.
+- Use Case 2: Geographic Coordinates: Represent geographic coordinates using records. Meaningful names for record classes and components contribute to the clarity of the code, allowing developers to quickly comprehend the structure and purpose of the geographic coordinate representation.
+
+## Ways to Master Java Records
```
Comment (Maiklins) [inline content/blog/2023/2023-12-03-use-cases-for-java-records.md:481]: This chapter only contains bullet lists. To create a nice reading flow, it'd be nice to have each bullet point list introduced by one or two sentences (like in the previous sections).

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+- Records automatically generate methods like equals(), hashCode(), and toString() for consistent behavior.
+- Best practices ensure effective use, including appropriate use cases and maintaining readability.
+
+### Next Steps We Should Take
+- **Apply the Concepts Learned:** Do not just stop at understanding — implement records in projects. Create immutable, clean, and readable data classes.
+- **Experiment and Refine:** Play with different scenarios and refine understanding. Experimentation leads to mastery.
+- **Share and Collaborate:** Share knowledge with peers, and collaborate on projects to learn from real-world use cases.
+
+### Further Reading and Resources We Could Refer To
+- **Official Java Documentation:** Dive deep into the [official documentation](https://docs.oracle.com/en/java/javase/14/index.html) to explore advanced topics and nuances.
```
Comment (Maiklins) [inline content/blog/2023/2023-12-03-use-cases-for-java-records.md:494]: A good in-depth read is also the JEP for records: https://openjdk.org/jeps/395

File: content/blog/2023/2023-12-03-use-cases-for-java-records.md (hunk 10 above)
Comment (sagaofsilence) [reply content/blog/2023/2023-12-03-use-cases-for-java-records.md:481]: Elaborated the points.

**Comment by sagaofsilence** [comment]
Requested changes done.

**Pull request merged**